Create guide for using an ESP devboard with USB_UART bridge chip as serial adapter

shvmm opened the pull request:
All functioning ESP devboards which can be flashed over USB are automagically also USB serial adapter boards.
The file `guides/devboard_as_flasher.rst` documents how to use one.

## Description:
![devboard-as-flasher-pi](https://github.com/user-attachments/assets/348303ae-f33e-4f38-986f-1be80d0e77cb)

![pi serial tty](https://github.com/user-attachments/assets/c93746d1-f0de-4228-9a2d-ff7d11fef8bb)

Sources:
https://pinout.xyz/
PNGs Credits:
https://freesvg.org/esp32-devkitc
https://freesvg.org/by/adlerweb
Idea from this video:
https://www.youtube.com/watch?v=ndEeFQ5mi_g

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome](https://github.com/esphome/esphome) with YAML changes (if applicable):** esphome/esphome#<esphome PR number goes here>

## Checklist:

  - [ ] I am merging into `next` because this is new documentation that has a matching pull-request in [esphome](https://github.com/esphome/esphome) as linked above.  
    or
  - [x] I am merging into `current` because this is a fix, change and/or adjustment in the current documentation and is not for a new component or feature.

  - [ ] Link added in `/index.rst` when creating new documents for new components or cookbook.

shvmm committed:
Create devboard_as_flasher.rst

need to add two example images to guides/images

shvmm committed:
Images part of PR4483

shvmm committed:
images part of PR4483

shvmm committed:
Update devboard_as_flasher.rst

shvmm committed:
Delete guides/images/devboard-as-flasher-pi.JPG

shvmm committed:
image part of PR4483

shvmm committed:
Update devboard_as_flasher.rst

shvmm committed:
Update devboard_as_flasher.rst

shvmm committed:
Update devboard_as_flasher.rst

clydebarrow commented on guides/devboard_as_flasher.rst:
```diff
@@ -0,0 +1,64 @@
+Using an ESP devboard as serial flasher
+=======================================
```
```suggestion
Using an ESP devboard as a USB-UART bridge
==========================================
```

clydebarrow commented on guides/devboard_as_flasher.rst:
```diff
@@ -0,0 +1,64 @@
+
+.. _devboard-as-flasher:
+
+If you have a functioning ESP devboard onto which you can flash firmware via USB port,
+it can be also used to flash/communicate via serial to other devices as well.
```
```suggestion
ESP development boards usually have an onboard USB interface, either built into the chip (e.g. ESP32-S3) or via a USB-UART bridge chip. However some ESP based devices not designed for development work don't bother with this, and only expose the UART0 pins (TX and RX) for flashing purposes.

Normally you would use a dedicated USB-UART interface board for this but what if you don't have one? In this "emergency" situation it is possible to use a development board that does have a USB-UART bridge chip to flash another device. This is achieved by holding the ESP chip in reset so that it doesn't interfere with the bridge chip operation.

It does NOT require any firmware to be flashed onto the development board and will not change anything already flashed onto it - it's purely a way to use the serial interface chip.

Note: most ESP32 S and C series boards do *not* have a separate USB-UART chip - they have it built into the ESP - so are not suitable for this application.
```

clydebarrow commented on guides/devboard_as_flasher.rst:
```diff
@@ -0,0 +1,64 @@
+
+.. note::
+
+    - On some boards like the AiThinker ESP32Cam, you additionally need to connect ``IO0`` with ``GND``
+      on the target device to force flash/download mode.
```
```suggestion
    - If the board has not previously had ESPHome loaded, it is probable that you will need to hold the ``IO0` pin low (i.e. connected to ground) to force the board into flash mode. This must be done before power is applied.
```

clydebarrow commented on guides/devboard_as_flasher.rst:
```diff
@@ -0,0 +1,64 @@
+
+    Connection diagram for an ESP flash target
+
+You need to make the following electrical connections:
+
```
```suggestion
You need to make the following electrical connections:

Note that the 5V connection on either board may be labelled either ``5V`` or ``VIN``. Some boards may not have a 5V connection and will require 3.3V only.

Rather than powering the target board from the flasher board, it is also possible to use a separate power supply, just make sure all the ground pins are connected together.
```

clydebarrow commented on guides/devboard_as_flasher.rst:
```diff
@@ -0,0 +1,64 @@
+      on the target device to force flash/download mode.
+    - Do not connect 3V3 to VIN of the target devices with a 3V3 LDO as it may lead to brownouts.
+
+.. collapse:: **Connecting to a Raspberry Pi UART**
+
+    Because it is really just exposing the onboard USB_UART bridge chip,
+    it can even be used to interface with serial console (using PuTTy or similar) on a RPi or any similar device.
+    You may need to enable serial interface from ``raspi-config`` first.
+    We don't decide what you may use it for.
+
+    The connection diagram is mostly unchanged.
+
+    - connect both ``EN`` and ``GND`` together in the flasher devboard
+    - connect ``GND`` to ``GND`` to prevent it from floating
+    - ``TX`` of flasher devboard to ``TX GPIO14`` of the ``RPi``
+    - ``RX`` of flasher devboard to ``RX GPIO15`` of the ``RPi``
+
+    .. figure:: /guides/images/devboard-as-flasher-pi.jpg
+        :align: center
+        :width: 75.0%
+
+        Connection diagram for a RPi serial target. Refer to https://pinout.xyz/
+
+    .. warning::
+
+        Powering a Raspberry Pi directly through the flasher devboard is not advised.
```
Is this section necessary? It seems too specific and may just cause confusion.

shvmm replied:
The plan is to expand this section to include debug instructions for RPico series boards. That's why
It is in a separate collapsible tho

clydebarrow commented on guides/devboard_as_flasher.rst:
```diff
@@ -0,0 +1,64 @@
+
+    - On some boards like the AiThinker ESP32Cam, you additionally need to connect ``IO0`` with ``GND``
+      on the target device to force flash/download mode.
+    - Do not connect 3V3 to VIN of the target devices with a 3V3 LDO as it may lead to brownouts.
```
```suggestion
    - Do not connect 3V3 to VIN of the target devices with a 3V3 LDO as it may lead to brownouts.
    
Once the connections are made, plug the flasher board into your computer via USB and proceed with flashing the target board via whichever means you intend to use.
```

clydebarrow reviewed (changes requested):
I have to say it took me a little time to realise what this guide was trying to show. Once the penny dropped that you are just using the USB-UART bridge chip it made more sense. So I've suggested additional clarification to make that clear.

I'd also suggest that the image showing the wiring connections has a poor choice for the target board - its image shows it *does* have a USB connector and a USB-UART chip, so there would be no need to use this technique to flash that board.

Finally, it would be worth adding a link to this in the `physical_device_connection` page, probably in the Materials table where it talks about a USB to serial adapter, since this what the dev board would be replacing.

clydebarrow commented on guides/devboard_as_flasher.rst:
```diff
@@ -0,0 +1,64 @@
+
+        Powering a Raspberry Pi directly through the flasher devboard is not advised.
+
+    Future Todo
+    -----------
+    
+    Verify functionality for UART debug ports on the following devices:
+        - Raspberry Pi Pico series
+        - Raspberry Pi 5
```
I would remove this. It's not useful to the reader.

shvmm replied:
Yeah, I'd move this artifact to the main PR description

shvmm committed:
Apply suggestions from code review

by clydebarrow

Co-authored-by: Clyde Stubbs <2366188+clydebarrow@users.noreply.github.com>

shvmm commented:
> I'd also suggest that the image showing the wiring connections has a poor choice for the target board - its image shows it _does_ have a USB connector and a USB-UART chip, so there would be no need to use this technique to flash that board.

It is easier for me to just remove the port, chip and accompanying mosfets from the target board.
The harder part would be finding a high quality svg for a target board with proper pinout labels.

shvmm committed:
Add files via upload

shvmm committed:
Update devboard_as_flasher.rst

shvmm committed:
fix lint tabs

shvmm committed:
Update devboard_as_flasher.rst

shvmm committed:
Update devboard_as_flasher.rst

shvmm committed:
Update physical_device_connection.rst

* Emphasis of data USB cables instead of power only cables

* Added references for flashing other devices using an ESP devboard

shvmm commented:
> Finally, it would be worth adding a link to this in the `physical_device_connection` page, probably in the Materials table where it talks about a USB to serial adapter, since this what the dev board would be replacing.

Added two links:
One in the section devoted to explaining the functionality of development boards 
And another in materials required section.

shvmm committed:
Update physical_device_connection.rst

added additional material against using power only usb cables for flashing ESPs

clydebarrow reviewed (approved):
LGTM
